python/pipeline/treadmill.py

```diff
@@ -95,4 +97,252 @@ def notify(self, key):
+        for n,running_period in enumerate(finalized_running_periods):
+            run_period_key = key.copy()
+            run_period_key['run_idx'] = n+1
+            run_period_key['run_onset'] = treadmill_times[running_period[0]]
+            run_period_key['run_offset'] = treadmill_times[running_period[-1]]
+            run_period_key['run_duration'] = run_period_key['run_offset'] - run_period_key['run_onset']
+            run_period_key['mean_velocity'] = np.nanmean(filt_treadmill_velocity[running_period])
+            run_period_key['mean_speed'] = np.nanmean(filt_treadmill_speed[running_period])
+            run_period_key['max_speed'] = np.nanmax(filt_treadmill_speed[running_period])
+            total_run_duration += run_period_key['run_duration']
+            running_period_keys.append(run_period_key)
+            
+        ## Calculate total running statistics
+        treadmill_duration = np.nanmax(treadmill_times) - np.nanmin(treadmill_times)
+        percent_running = total_run_duration / treadmill_duration * 100
+        
+        ## Insert key values 
+        running_key = {**key, 'processed_treadmill': filt_treadmill_speed, 'total_run_duration': total_run_duration, 
+                       'percent_running': percent_running, 'mean_run_velocity': mean_run_velocity,
+                       'mean_run_speed': mean_run_speed, 'max_run_speed': max_run_speed, 
+                       'num_run_periods': num_run_periods}
+        self.insert1(running_key)
+        self.Period.insert(running_period_keys)
+
+        
+    def _combine_running_periods(self, running_indices, tread_times, maximum_time_gap):
+
+        """
+        Given a flattened 1d array/list of indices, combines defined indices and indices between them if their gap
+        is less than maximum_time_gap.
+
+            Parameters:
+                running_indices: Flattened list or numpy array of indices corresponding to detected running times
+                tread_times: List or numpy array of times each recording on the treadmill was taken
+                maximum_time_gap: Numeric value of maximum gap before running periods should not be combined. 
+                                  Value is written in seconds.
+
+            Returns:
+                combined_running_fragments: List of lists of continuous incrementing indices corresponding to
+                                            single periods of running.
+        """
+
+        ## Splits list of indices into a list of lists. Each sublist contains indices which continuously increase by 1.
+        ## This function also drops sublists which only contain a single idx.
+        running_fragments = clocktools.find_idx_boundaries(running_indices, drop_single_idx=True)
+
+        ## If we have two or more fragments, test if any should be combined
+        if len(running_fragments) > 1:
+
+            ## Basic idea: Have a temporary start index. Loop through all lists of indices. Whenever there is a list
```

Comment:
I'm confused about what this is doing. So, if there is a long enough gap between running periods then they are combined? Shouldn't it be the other way around, i.e. if the gap is short enough then combine them?

Reply:
You can think of it like having two cursors, one for start_idx and another for end_idx. Start by setting start and end cursors on the start and end of the first fragment. When looking at the second fragment, one of two things happens:
1. The gap between the first and second fragment is below threshold. In that case, set the end cursor to the end of the second fragment and keep looping through.
2. The gap between the first and second fragment is greater than the threshold. In that case store an np.arange from the start cursor to the end cursor before setting those cursors to the start and end of the second fragment, then continue looping through fragments.

Does that seem reasonable to you?